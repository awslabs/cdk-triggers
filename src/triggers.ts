import { join } from 'path';
import * as lambda from '@aws-cdk/aws-lambda';
import { Construct, CustomResource, CustomResourceProvider, CustomResourceProviderRuntime } from '@aws-cdk/core';

export interface AfterCreateProps {
  /**
   * Resources to trigger on. Resources can come from any stack in the app.
   *
   * @default [] Run the trigger at any time during stack deployment.
   */
  readonly resources?: Construct[];

  /**
   * The handler to execute once after all the resources are created.
   *
   * The trigger will be executed every time the handler changes (code or
   * configuration).
   */
  readonly handler: lambda.Function;
}

export class AfterCreate extends Construct {
  constructor(scope: Construct, id: string, props: AfterCreateProps) {
    super(scope, id);

    const provider = CustomResourceProvider.getOrCreateProvider(this, 'AWSCDK.TriggerCustomResourceProvider', {
      runtime: CustomResourceProviderRuntime.NODEJS_14_X,
      codeDirectory: join(__dirname, 'handler'),
      policyStatements: [
        {
          Effect: 'Allow',
          Action: ['lambda:InvokeFunction'],
          Resource: ['*'], // TODO?
        },
      ],
    });

    const resource = new CustomResource(this, 'Resource', {
      resourceType: 'Custom::Trigger',
      serviceToken: provider.serviceToken,
      properties: {
        // we use 'currentVersion' because a new version is created every time the
        // handler changes (either code or configuration). this will have the effect
        // that the trigger will be executed every time the handler is changed.
        HandlerArn: props.handler.currentVersion.functionArn,
      },
    });

    // add a dependency between our resource and the resources we want to invoke
    // after.
    resource.node.addDependency(...props.resources ?? []);
  }
}