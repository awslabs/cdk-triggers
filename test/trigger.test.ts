import { ResourcePart, SynthUtils } from '@aws-cdk/assert';
import '@aws-cdk/assert/jest';
import * as lambda from '@aws-cdk/aws-lambda';
import * as sns from '@aws-cdk/aws-sns';
import { Stack } from '@aws-cdk/core';
import * as triggers from '../src';

test('minimal usage', () => {
  // GIVEN
  const stack = new Stack();
  const triggeringResource = new sns.Topic(stack, 'MyTopic');
  const trigger = new lambda.Function(stack, 'MyTriggerHandler', {
    runtime: lambda.Runtime.NODEJS_12_X,
    code: lambda.Code.fromInline('zoo'),
    handler: 'index.handler',
  });

  // WHEN
  new triggers.AfterCreate(stack, 'MyTrigger', {
    handler: trigger,
    resources: [triggeringResource],
  });

  // THEN
  expect(SynthUtils.synthesize(stack).template).toMatchSnapshot();

  expect(stack).toHaveResource('Custom::Trigger', {
    HandlerArn: {
      Ref: 'MyTriggerHandlerCurrentVersionC0B6BBD40f3abd954eb77fda7e548d681c7fa667',
    },
  });

  expect(stack).toHaveResource('Custom::Trigger', {
    DependsOn: ['MyTopic86869434'],
  }, ResourcePart.CompleteDefinition);
});
