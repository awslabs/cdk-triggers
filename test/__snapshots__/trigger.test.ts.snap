// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`minimal usage 1`] = `
Object {
  "Parameters": Object {
    "AssetParameters9c7118384671661d984b2df05cd6022e0285691ee053fa0b93120726717a0f84ArtifactHash9AEC0859": Object {
      "Description": "Artifact hash for asset \\"9c7118384671661d984b2df05cd6022e0285691ee053fa0b93120726717a0f84\\"",
      "Type": "String",
    },
    "AssetParameters9c7118384671661d984b2df05cd6022e0285691ee053fa0b93120726717a0f84S3BucketFA370235": Object {
      "Description": "S3 bucket for asset \\"9c7118384671661d984b2df05cd6022e0285691ee053fa0b93120726717a0f84\\"",
      "Type": "String",
    },
    "AssetParameters9c7118384671661d984b2df05cd6022e0285691ee053fa0b93120726717a0f84S3VersionKeyBAA707E6": Object {
      "Description": "S3 key for asset version \\"9c7118384671661d984b2df05cd6022e0285691ee053fa0b93120726717a0f84\\"",
      "Type": "String",
    },
  },
  "Resources": Object {
    "AWSCDKTriggerCustomResourceProviderCustomResourceProviderHandler97BECD91": Object {
      "DependsOn": Array [
        "AWSCDKTriggerCustomResourceProviderCustomResourceProviderRoleE18FAF0A",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "AssetParameters9c7118384671661d984b2df05cd6022e0285691ee053fa0b93120726717a0f84S3BucketFA370235",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                Object {
                  "Fn::Select": Array [
                    0,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters9c7118384671661d984b2df05cd6022e0285691ee053fa0b93120726717a0f84S3VersionKeyBAA707E6",
                        },
                      ],
                    },
                  ],
                },
                Object {
                  "Fn::Select": Array [
                    1,
                    Object {
                      "Fn::Split": Array [
                        "||",
                        Object {
                          "Ref": "AssetParameters9c7118384671661d984b2df05cd6022e0285691ee053fa0b93120726717a0f84S3VersionKeyBAA707E6",
                        },
                      ],
                    },
                  ],
                },
              ],
            ],
          },
        },
        "Handler": "__entrypoint__.handler",
        "MemorySize": 128,
        "Role": Object {
          "Fn::GetAtt": Array [
            "AWSCDKTriggerCustomResourceProviderCustomResourceProviderRoleE18FAF0A",
            "Arn",
          ],
        },
        "Runtime": "nodejs14.x",
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "AWSCDKTriggerCustomResourceProviderCustomResourceProviderRoleE18FAF0A": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Sub": "arn:\${AWS::Partition}:iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
          },
        ],
        "Policies": Array [
          Object {
            "PolicyDocument": Object {
              "Statement": Array [
                Object {
                  "Action": Array [
                    "lambda:InvokeFunction",
                  ],
                  "Effect": "Allow",
                  "Resource": Array [
                    "*",
                  ],
                },
              ],
              "Version": "2012-10-17",
            },
            "PolicyName": "Inline",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "MyTopic86869434": Object {
      "Type": "AWS::SNS::Topic",
    },
    "MyTriggerB6CCCACE": Object {
      "DeletionPolicy": "Delete",
      "DependsOn": Array [
        "MyTopic86869434",
      ],
      "Properties": Object {
        "HandlerArn": Object {
          "Ref": "MyTriggerHandlerCurrentVersionC0B6BBD40f3abd954eb77fda7e548d681c7fa667",
        },
        "ServiceToken": Object {
          "Fn::GetAtt": Array [
            "AWSCDKTriggerCustomResourceProviderCustomResourceProviderHandler97BECD91",
            "Arn",
          ],
        },
      },
      "Type": "Custom::Trigger",
      "UpdateReplacePolicy": "Delete",
    },
    "MyTriggerHandlerCurrentVersionC0B6BBD40f3abd954eb77fda7e548d681c7fa667": Object {
      "Properties": Object {
        "FunctionName": Object {
          "Ref": "MyTriggerHandlerD6B1FF23",
        },
      },
      "Type": "AWS::Lambda::Version",
    },
    "MyTriggerHandlerD6B1FF23": Object {
      "DependsOn": Array [
        "MyTriggerHandlerServiceRoleFC0CFFAB",
      ],
      "Properties": Object {
        "Code": Object {
          "ZipFile": "zoo",
        },
        "Handler": "index.handler",
        "Role": Object {
          "Fn::GetAtt": Array [
            "MyTriggerHandlerServiceRoleFC0CFFAB",
            "Arn",
          ],
        },
        "Runtime": "nodejs12.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "MyTriggerHandlerServiceRoleFC0CFFAB": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
  },
}
`;
